framework/src/positions/Positions.C

```diff
@@ -0,0 +1,244 @@
+{
+  // Unroll in every dimension available
+  std::vector<Point> temp_2d_unrolled;
+  std::vector<Point> temp_3d_unrolled;
+  std::vector<Point> temp_4d_unrolled;
+  for (auto vec : _positions_2d)
+    temp_2d_unrolled.insert(temp_2d_unrolled.end(), vec.begin(), vec.end());
+  for (auto vec_vec : _positions_3d)
+    for (auto vec : vec_vec)
+      temp_3d_unrolled.insert(temp_3d_unrolled.end(), vec.begin(), vec.end());
+  for (auto vec_vec_vec : _positions_4d)
+    for (auto vec_vec : vec_vec_vec)
+      for (auto vec : vec_vec)
+        temp_4d_unrolled.insert(temp_4d_unrolled.end(), vec.begin(), vec.end());
+
+  // for now we wont even tolerate a different ordering
+  // 2D & 1D: check for conflicts
+  if (temp_2d_unrolled.size() && _positions.size() && temp_2d_unrolled != _positions)
+    mooseError("Inconsistency between the 2D and 1D position vectors detected. "
+               "The 2D positions must unroll into the 1D positions");
+
+  // 3D vs 2D & 1D
+  if (temp_3d_unrolled.size() && temp_2d_unrolled.size() && temp_3d_unrolled != temp_3d_unrolled)
+    mooseError("Inconsistency between the 3D and 2D position vectors detected. "
+               "The 3D positions must unroll the same way as the 2D positions");
+  if (temp_3d_unrolled.size() && _positions.size() && temp_3d_unrolled != _positions)
+    mooseError("Inconsistency between the 3D and 1D position vectors detected. "
+               "The 3D positions must unroll into the 1D positions");
+
+  // 4D vs all lower dimensions
+  if (temp_4d_unrolled.size() && temp_3d_unrolled.size() && temp_4d_unrolled != temp_3d_unrolled)
+    mooseError("Inconsistency between the 4D and 3D position vectors detected. "
+               "The 4D positions must unroll the same way as the 3D positions");
+  if (temp_4d_unrolled.size() && temp_2d_unrolled.size() && temp_4d_unrolled != temp_2d_unrolled)
+    mooseError("Inconsistency between the 4D and 2D position vectors detected. "
+               "The 4D positions must unroll the same way as the 2D positions");
+  if (temp_4d_unrolled.size() && _positions.size() && temp_4d_unrolled != _positions)
+    mooseError("Inconsistency between the 4D and 1D position vectors detected. "
+               "The 4D positions must unroll into the 1D positions");
+
+  // Use any of the higher D to set the one D vector
+  if (!_positions.size())
+  {
+    if (temp_2d_unrolled.size())
+      _positions = temp_2d_unrolled;
+    else if (temp_3d_unrolled.size())
+      _positions = temp_3d_unrolled;
+    else if (temp_4d_unrolled.size())
+      _positions = temp_4d_unrolled;
+    else
+      mooseError("Positions::unrollMultiDPositions() may only be called if there is at least one "
+                 "non-empty positions vector.");
+  }
+}
+
+void
+Positions::finalize()
+{
+  // Gather up the positions vector on all ranks
+
+  if (_need_broadcast)
+  {
+    // TODO: be able to do this
+    // _communicator.allgather(_positions, /* identical buffer lengths = */ false);
```

Comment:
Why can't we do this? If you include `libmesh/parallel_algebra.h` I feel like this should work

Reply:
indeed